sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -136,3 +141,53 @@ private[sql] case class SparkUserDefinedFunction(
     }
   }
 }
+
+/**
+ * A User Defined Function for wrapping an [[Aggregator]] in a form consumable by Data Frames.
+ * Can be either instantiated directly or via registration, for example:
+ * {{{
+ *   val agg: Aggregator[IN, BUF, OUT] = // typed aggregator
+ *   val udaf1 = UserDefinedAggregator(agg)
+ *   val udaf2 = spark.udf.registerAggregator("agg", agg)
+ * }}}
+ *
+ * @since 3.0.0
+ */
+@Experimental
+case class UserDefinedAggregator[IN: TypeTag, BUF, OUT](
+    aggregator: Aggregator[IN, BUF, OUT],
+    name: Option[String] = None,
+    nullable: Boolean = true,
+    deterministic: Boolean = true) extends UserDefinedFunction {
+
+  @scala.annotation.varargs
+  def apply(exprs: Column*): Column = {
+    val inputEncoder = (ExpressionEncoder[IN]()).resolveAndBind()
```

Comment:
nit: I think it's better to put the input encoder in the constructor, so that we don't need to require TypeTag for this class.